### ISSUE TITLE
Not able to unzip file to sdcard

### ISSUE THREAD

**Issue by abhishekmittal4007** [TeamAmaze/AmazeFileManager#845]
Hello,

I want to extract the zip file (~858 Mb) to sdcard using SAF. For that, I have already taken the permission and saved the root uri in my shared preference. After going through your code, I took the getDcumentFile() method from your FileUtils class. Everything is going well but the problem is, extraction process is taking approx. 1hr to extract this file to sdcard.
I am attaching my extraction code here. Please let me know if there is anything wrong.

public void start() {
        destinationPath = PreferenceManager.getDefaultSharedPreferences(context).getString("PATH", null);
        treeUri = Uri.parse(PreferenceManager.getDefaultSharedPreferences(context).getString("URI", null));
        DocumentFile pickedDir = DocumentFile.fromTreeUri(context, treeUri);
        zipd = pickedDir.findFile("POSexternal.zip");
        unzipToSDcard(zipd);
    }

    /*
    for extracting zip to sdcard or mount storage
     */
    public boolean unzipToSDcard(DocumentFile zipFile/*, File destinationDir*/) {
        ZipFile zip = null;
        try {
            String path = SDCardUtil.getRealPathFromURI(context, zipFile.getUri());
            zip = new ZipFile(path);
            for (Enumeration e = zip.entries(); e.hasMoreElements(); ) {
                // adding all the elements to be extracted to an array list
                ZipEntry entry = (ZipEntry) e.nextElement();
                ttl_size += entry.getSize();
                arrayList.add(entry);
            }
            for (ZipEntry z_entry : arrayList) {
                unzipEntry(zip, z_entry, destinationPath);
            }
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        } finally {
            if (zip != null) {
                try {
                    zip.close();
                } catch (IOException ignored) {
                    ignored.printStackTrace();
                }
            }
        }
        return true;
    }

    private void unzipEntry(ZipFile zipfile, ZipEntry entry, String outputDir) throws Exception {
        if (entry.isDirectory()) {
            getDocumentFile(new File(outputDir, "/" + entry.getName()), true, context);
            return;
        }
        File outputFile = new File(outputDir, entry.getName());
        if (!outputFile.getParentFile().exists()) {
            getDocumentFile(outputFile.getParentFile(), true, context);
        }
        DocumentFile df = getDocumentFile(outputFile, false, context);
        InputStream inputStream = zipfile.getInputStream(entry);
        OutputStream outputStream = context.getContentResolver().openOutputStream(df.getUri());
        try {
            int len = 0;
            byte buf[] = new byte[GenericCopyUtil.DEFAULT_BUFFER_SIZE];
            while ((len = inputStream.read(buf, 0, buf.length)) > 0) {
                outputStream.write(buf, 0, len);
                totalBytesCopied += len;
                int progress = (int) Math.round(((double) totalBytesCopied / (double) ttl_size) * 100);
                publishProgress(entry.getName(), "" + progress);
            }
        } catch (IOException e) {
            e.printStackTrace();
            outputStream.close();
        } finally {
//            outputStream.flush();
//            outputStream.close();
//            inputStream.close();
        }
    }

    public static DocumentFile getDocumentFile(final File file, final boolean isDirectory, Context context) {
        String baseFolder = getExtSdCardFolder(file, context);
//        boolean originalDirectory = false;
        if (baseFolder == null) {
            return null;
        }

        String relativePath = null;
        try {
            String fullPath = file.getCanonicalPath();
            if (!baseFolder.equals(fullPath))
                relativePath = fullPath.substring(baseFolder.length() + 1);
        } catch (IOException e) {
            return null;
        } catch (Exception f) {
            //continue
        }

        String as = PreferenceManager.getDefaultSharedPreferences(context).getString("URI", null);

        Uri treeUri = null;
        if (as != null) treeUri = Uri.parse(as);
        if (treeUri == null) {
            return null;
        }

        // start with root of SD card and then parse through document tree.
        DocumentFile document = DocumentFile.fromTreeUri(context, treeUri);
//        if (originalDirectory) return document;
        Log.d("path:::", relativePath);
        String[] parts = relativePath.split("\\/");
        for (int i = 0; i < parts.length; i++) {
            DocumentFile nextDocument = document.findFile(parts[i]);
            if (nextDocument == null) {
                if ((i < parts.length - 1) || isDirectory) {
                    nextDocument = document.createDirectory(parts[i]);
                } else {
                    nextDocument = document.createFile("image", parts[i]);
                }
            }
            document = nextDocument;
        }
        return document;
    }

Thank You

**Comment by EmmanuelMess** [comment]
So, what's the issue? Is Amaze taking too long (more than normal)?

**Comment by abhishekmittal4007** [comment]
No, I m not using amaze at all. The thing is, for unzipping the zip file (dealing with DocumentFile), I have used a part of your code, but it's taking to long to unzip.
I m just asking to see my piece of code that I have attached with this ticket and let me know the error that why is it taking too long to extract to sdcard.

[code.txt](https://github.com/TeamAmaze/AmazeFileManager/files/1508582/code.txt)

Thank You

**Comment by EmmanuelMess** [comment]
Go to StackOverflow.com

**Comment by abhishekmittal4007** [comment]
After searching a lot on stackoverflow...I have asked you. But its so disappointing that having no clue for this problem, you have labeled this problem as "closed".

**Comment by EmmanuelMess** [comment]
This is for issues relating the app, if you want answers about your code ask a question on [stackoverflow.com](stackoverflow.com).

**Comment by prathamApp** [comment]
Yeah I know that, But the thing is I have used your ExtractService.java class in service package and my code is similar to yours but the thing is your's work fast but mine it is taking ~1hr . Any guess why...?
